gx_suite_definitions/proteomics_srm.ipynb

```diff
@@ -0,0 +1,276 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import synapseclient\n",
+    "\n",
+    "import great_expectations as gx\n",
+    "\n",
+    "context = gx.get_context(project_root_dir='../src/agoradatatools/great_expectations')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Create Expectation Suite for Proteomics SRM Data"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Get Example Data File"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "syn = synapseclient.Synapse()\n",
+    "syn.login()\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "proteomics_srm_data_file = syn.get(\"syn52579910\").path"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Create Validator Object on Data File"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "validator = context.sources.pandas_default.read_json(\n",
+    "    proteomics_srm_data_file\n",
+    ")\n",
+    "validator.expectation_suite_name = \"proteomics_srm\"\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Add Expectations to Validator Object For Each Column"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# uniqid\n",
+    "validator.expect_column_values_to_be_of_type(\"uniqid\", \"str\")\n",
+    "validator.expect_column_values_to_not_be_null(\"uniqid\")\n",
+    "validator.expect_column_values_to_match_regex(\"uniqid\", \"^(?:[A-Za-z0-9_.-]+\\\\|[A-Za-z0-9_.-]+|[OPQ][0-9][A-Z0-9]{3}[0-9]|[A-NR-Z][0-9](?:[A-Z][A-Z0-9]{2}[0-9]){1,2})$\")"
+   ]
```

Comment:
I'm not a regex wizard, so I dropped this regex into regexr.com:
`^(?:[A-Za-z0-9_.-]+\\\\|[A-Za-z0-9_.-]+|[OPQ][0-9][A-Z0-9]{3}[0-9]|[A-NR-Z][0-9](?:[A-Z][A-Z0-9]{2}[0-9]){1,2})$`
 
And it didn't match a valid uniqid: `CLTB|P09497`

I had to modify the `hgnc|` portion of the regex to get it to match:
`^(?:[A-Za-z0-9_.-]+\|[OPQ][0-9][A-Z0-9]{3}[0-9]|[A-NR-Z][0-9](?:[A-Z][A-Z0-9]{2}[0-9]){1,2})$`

It looks like part of the hgnc regex got duplicated, and the literal pipe character isn't properly escaped...

The other proteomics PR has the same regex, so if we need to update this we'll need to update that one too...

Reply:
The GX report you linked indicates that the regexes all match as expected (everything passed), but the regexes they report as matching are not the correct ones. For uniprot, it says it applied the hgnc regex, and for uniqid it only applied the 2nd half of the regex. Any idea what is going on?

![Screen Shot 2024-08-29 at 1 02 03 PM](https://github.com/user-attachments/assets/537b0f2d-0f7d-472f-8204-af1a867f38a7)

Reply:
Finally getting back to this. Sorry for the delay! I will look into this and see what I can find out.
/

Reply:
Hmm, I'm not sure what was going on with the rendering when you viewed it, but when I look at the report linked above I see the correct regexes being applied. I also re-ran it to be sure and it came out correct again.

![Screenshot 2024-09-16 at 11 39 34 AM](https://github.com/user-attachments/assets/da539ac0-0cf4-4262-8ac7-80e60058f229)

Thanks for catching the duplicate `hgnc_symbol` duplicate regex. I updated the string to no longer have that in there. As for the pipe escaping, there is a slight difference between using regexr.com and the way that the GX regex expectation works. I think because the GX rule is being parsed into JSON we have to double escape the pipe, so this `^(?:[A-Za-z0-9_.-]+\|[OPQ][0-9][A-Z0-9]{3}[0-9]|[A-NR-Z][0-9](?:[A-Z][A-Z0-9]{2}[0-9]){1,2})$` in regexr is equivalent to this `^(?:[A-Za-z0-9_.-]+\\|[OPQ][0-9][A-Z0-9]{3}[0-9]|[A-NR-Z][0-9](?:[A-Z][A-Z0-9]{2}[0-9]){1,2})$` in our GX suite.

Reply:
Hmm, I'm not sure what was going on with the first file I linked, but when I re-ran it came out [correct](https://www.synapse.org/Synapse:syn61810955).

![Screenshot 2024-09-16 at 11 39 34 AM](https://github.com/user-attachments/assets/da539ac0-0cf4-4262-8ac7-80e60058f229)

Thanks for catching the duplicate `hgnc_symbol` duplicate regex. I updated the string to no longer have that in there. As for the pipe escaping, there is a slight difference between using regexr.com and the way that the GX regex expectation works. I think because the GX rule is being parsed into JSON we have to double escape the pipe, so this `^(?:[A-Za-z0-9_.-]+\|[OPQ][0-9][A-Z0-9]{3}[0-9]|[A-NR-Z][0-9](?:[A-Z][A-Z0-9]{2}[0-9]){1,2})$` in regexr is equivalent to this `^(?:[A-Za-z0-9_.-]+\\|[OPQ][0-9][A-Z0-9]{3}[0-9]|[A-NR-Z][0-9](?:[A-Z][A-Z0-9]{2}[0-9]){1,2})$` in our GX suite.

Reply:
Oh, wait I'm realizing that this wasn't a duplicate `hgnc_symbol` but rather an additional matching scheme that was necessary for some of the records in the `proteomics` dataset. I think what I have done here is good for this dataset though.

Reply:
The report looks good now, that other one was probably from an old build or something...

I just aligned the three regex strings, and I see one unexpected difference in the 2nd half (unprot part) of the uniquid regex:

uniprot has `[A-Z]`:
`^[OPQ][0-9][A-Z0-9]{3}[0-9]|[A-NR-Z][0-9]([A-Z][A-Z0-9]{2}[0-9]){1,2}$\`

uniqID has `?:[A-Z]`
`^(?:[A-Za-z0-9_.-]+\\\\|[OPQ][0-9][A-Z0-9]{3}[0-9]|[A-NR-Z][0-9](?:[A-Z][A-Z0-9]{2}[0-9]){1,2})$\`

The uniprot regex provided by uniprot was this one:

`[OPQ][0-9][A-Z0-9]{3}[0-9]|[A-NR-Z][0-9]([A-Z][A-Z0-9]{2}[0-9]){1,2}`

Did we add the `?:` to convert this to a non-capturing group on purpose?

Reply:
To be honest, I don't remember why I had the `?:` in the string, but it seems to work without it so I have removed it.